ftwhttp/header.go

```diff
@@ -4,149 +4,182 @@
+	h.entries = append(h.entries, HeaderTuple{name, value})
+}
 
-	sorted := h.getSortedHeadersByName()
-
-	for _, key := range sorted {
-		// we want all headers "as-is"
-		s := key + ": " + h[key] + "\r\n"
-		if _, err := ws.WriteString(s); err != nil {
-			return err
+// Set replaces any existing HTTP headers of the same canonical
+// name with this new entry.
+func (h *Header) Set(name string, value string) {
+	key := canonicalKey(name)
+	retainees := []HeaderTuple{}
+	for _, tuple := range h.entries {
+		if canonicalKey(tuple.Name) != key {
+			retainees = append(retainees, tuple)
 		}
 	}
+	h.entries = retainees
+	h.Add(name, value)
+}
 
-	return nil
+// Returns true if the Header contains any HTTP header that
+// matches the canonical name.
+func (h *Header) HasAny(name string) bool {
+	key := canonicalKey(name)
+	_, ok := h.canonicalNames[key]
+	return ok
+}
 
+// Returns true if the Header contains any HTTP header that
+// matches the canonical name and canoncial value.
+// Values are compared using strings.EqualFold.
+func (h *Header) HasAnyValue(name string, value string) bool {
+	identity := func(a string) string { return a }
+	return h.hasAnyValue(name, value, identity, identity, strings.EqualFold)
 }
 
-// WriteBytes writes a header in a ByteWriter.
-func (h Header) WriteBytes(b *bytes.Buffer) (int, error) {
-	sorted := h.getSortedHeadersByName()
-	count := 0
-	for _, key := range sorted {
-		// we want all headers "as-is"
-		s := key + ": " + h[key] + "\r\n"
-		log.Trace().Msgf("Writing header: %s", s)
-		n, err := b.Write([]byte(s))
-		count += n
-		if err != nil {
-			return count, err
-		}
-	}
+// Returns true if the Header contains any HTTP header that
+// matches the canonical name and has a value containing the
+// specified substring.
+// Both, the header value and the search string are lower-cased
+// before performing the search.
+func (h *Header) HasAnyValueContaining(name string, value string) bool {
+	return h.hasAnyValue(name, value, strings.ToLower, strings.ToLower, strings.Contains)
+}
 
-	return count, nil
+// Returns all HeaderTuples that match the canonical header name.
+// If no matches are found the returned array will be empty.
+func (h *Header) GetAll(name string) []HeaderTuple {
+	return h.getAll(canonicalKey(name), canonicalKey)
 }
 
-// Clone returns a copy of h
-func (h Header) Clone() Header {
-	clone := make(Header)
+// Write writes the header to the provided writer
+func (h *Header) Write(writer io.Writer) error {
+	buf := bufio.NewWriter(writer)
+	for _, tuple := range h.entries {
+		if log.Trace().Enabled() {
+			log.Trace().Msgf("Writing header: %s: %s", tuple.Name, tuple.Value)
+		}
+		if _, err := buf.WriteString(tuple.Name); err != nil {
+			return err
+		}
+		if _, err := buf.WriteString(headerSeparator); err != nil {
+			return err
+		}
+		if _, err := buf.WriteString(tuple.Value); err != nil {
+			return err
+		}
+		if _, err := buf.WriteString(headerDelimiter); err != nil {
+			return err
+		}
```

Comment:
Does it add value doing it like this? Why not all four at once, and return the error?

Reply:
This is a tiny bit faster, according to the benchmark. Probably not worth it though, you're right.